--- IDT_Latest/SPO2.h	Thu Mar 14 11:50:56 2019
+++ SPO2/SPO2.h	Thu Mar 14 15:40:31 2019
@@ -4 +4,6 @@
-#include "mbed.h"
+#if defined(__CA78K0R__) || defined(__CCRL__) || defined(__ICCRL78__) 
+#include "r_cg_macrodriver.h"
+#include "r_cg_iica.h"
+#include "r_cg_userdefine.h"
+#endif //#if defined(__CA78K0R__) || defined(__CCRL__) || defined(__ICCRL78__) 
+
@@ -58 +63 @@ class SPO2
-    void dither(int16_t *x, uint16_t len, uint16_t offset, const float *rel_vals, uint16_t num_vals, int32_t *correls, float *offsets);
+    void dither(int16_t *x, uint16_t len, uint16_t offset, const uint16_t *rel_vals, uint16_t num_vals, int32_t *correls, uint16_t *offsets);
